### PR TITLE
Pokémon R/B: Migrate support into Bizhawk Client

### PR DESCRIPTION
**Pull request by Alchav**
## What is this fixing or adding?
- Removes the Pokémon Client, adding support for Red and Blue to the Bizhawk Client.
- Adds `/bank` commands that mirror SDV's, allowing transferring money into and out of the EnergyLink storage.
- Adds a fix to the base patch so that the progressive card key counter will not increment beyond 10, which would lead to receiving glitch items. This value is checked against and verified that it is not > 10 as part of crash detection by the client, to prevent erroneous location checks when the game crashes, so this is relevant to the new client (although shouldn't happen unless you're using !getitem, or putting progressive card keys as item link replacement items)

## How was this tested?
By testing it. A client has also been made available in #pkmn-red-blue for further testing by others.

### CODE REVIEW

**Review by Zunawe — COMMENTED** [review]
Lots of single quoted strings that should be double quoted.

If this is a supported way to dynamically add or remove commands
```py
ctx.command_processor.commands["bank"] = cmd_bank
```
then there's probably a reasonable interface that could be added to BizHawkClient where you define your commands on a classvar and the main loop will add and remove your commands when it detects ROM changes.

As it is though, I don't think CommonClient is built in a way that expects dynamically added commands, so it'd be better to start from there and work down. I think this method is fine for now, since it will only be present if you switch games after playing R/B and try to use it, and it checks you and tells you what you did wrong.

**Comment by Zunawe** [comment]
Only just noticed this was a draft :sweat_smile:. Didn't mean to be over-eager. The discord feed didn't show draft at all, so I just clicked it and looked around.

File: inno_setup.iss [hunk 1]
```diff
@@ -32,7 +32,7 @@ AllowNoIcons=yes
 SetupIconFile={#MyAppIcon}
 UninstallDisplayIcon={app}\{#MyAppExeName}
 ; you will likely have to remove the following signtool line when testing/debugging locally. Don't include that change in PRs.
-SignTool= signtool
+;SignTool= signtool
```
[the remark this answers is not in the document]
Comment (Alchav) [reply inno_setup.iss:35]: Of course

File: worlds/pokemon_rb/client.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+
+    def on_package(self, ctx, cmd, args):
+        if cmd == 'Connected':
+            if 'death_link' in args['slot_data'] and args['slot_data']['death_link']:
+                self.set_deathlink = True
+                self.last_death_link = time.time()
+            ctx.set_notify(f"EnergyLink{ctx.team}")
+        elif cmd == "Retrieved":
+            if f"EnergyLink{ctx.team}" in args["keys"]:
+                ctx.current_energy_link_value = args["keys"][f"EnergyLink{ctx.team}"]
```
[the remark this answers is not in the document]
Comment (Alchav) [reply worlds/pokemon_rb/client.py:244]: What do you mean?

File: worlds/pokemon_rb/client.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+            await ctx.send_msgs([{"cmd": "LocationScouts", "locations": hints, "create_as_hint": 2}])
+        self.auto_hints.update(hints)
+
+        # DEATHLINK
+
+        if "DeathLink" in ctx.tags:
+            if data["Deathlink"][0] == 3:
+                await ctx.send_death(ctx.player_names[ctx.slot] + " is out of usable Pokémon! "
+                                     + ctx.player_names[ctx.slot] + " blacked out!")
+                await write(ctx.bizhawk_ctx, [(DATA_LOCATIONS["Deathlink"][0], [0], "WRAM")])
```
[the remark this answers is not in the document]
Comment (Alchav) [reply worlds/pokemon_rb/client.py:190]: Writing something into RAM during a crash isn't going to make a difference. I just don't want to read anything.

File: worlds/pokemon_rb/client.py [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+                await write(ctx.bizhawk_ctx, [(DATA_LOCATIONS["Deathlink"][0], [1], "WRAM")])
+
+        # BANK
+
+        if self.banking_command:
+            original_money = await read(ctx.bizhawk_ctx, [(0x141F, 3, "WRAM")])
+            original_money = original_money[0]
+            original_money = [int(byte) for byte in original_money]
+            money = [int(hex(byte).split("x")[1]) for byte in original_money]
+            money = (money[0] * 10000) + (money[1] * 100) + money[2]
```
[the remark this answers is not in the document]
Comment (Alchav) [reply worlds/pokemon_rb/client.py:203]: It's binary-coded decimal, where 0x99 means 99.

File: worlds/pokemon_rb/client.py [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+
+    async def validate_rom(self, ctx):
+        game_name = await read(ctx.bizhawk_ctx, [(0x134, 12, "ROM")])
+        game_name = game_name[0].decode("ascii")
+        if game_name in ("POKEMON RED\00", "POKEMON BLUE"):
+            ctx.game = self.game
+            ctx.items_handling = 0b001
+            ctx.command_processor.commands["bank"] = cmd_bank
+            seed_name = await read(ctx.bizhawk_ctx, [(0xFFDB, 21, "ROM")])
+            ctx.seed_name = seed_name[0].split(bytes([0]))[0].decode("ascii")
```
[the remark this answers is not in the document]
Comment (Alchav) [reply worlds/pokemon_rb/client.py:75]: Done

File: worlds/LauncherComponents.py [hunk 6]
```diff
@@ -89,6 +89,9 @@ def launch_textclient():
     Component('SNI Client', 'SNIClient',
               file_identifier=SuffixIdentifier('.apz3', '.apm3', '.apsoe', '.aplttp', '.apsm', '.apsmz3', '.apdkc3',
                                                '.apsmw', '.apl2ac')),
+    # BizHawk
+    Component("BizHawk Client", "BizHawkClient", component_type=Type.CLIENT,
+              file_identifier=SuffixIdentifier()),
```
[the remark this answers is not in the document]
Comment (Alchav) [reply worlds/LauncherComponents.py:94]: Removed

File: worlds/pokemon_rb/client.py [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+                        "cmd": "Set", "key": f"EnergyLink{ctx.team}", "operations":
+                            [{"operation": "add", "value": self.banking_command * BANK_EXCHANGE_RATE},
+                             {"operation": "max", "value": 0}],
+                    }])
+            self.banking_command = None
+
+        # VICTORY
+
+        if data["EventFlag"][280] & 1 and not ctx.finished_game:
+            await ctx.send_msgs([{"cmd": "StatusUpdate", "status": 30}])
```
[the remark this answers is not in the document]
Comment (Alchav) [reply worlds/pokemon_rb/client.py]: Done

**Comment by Alchav** [comment]
> Only just noticed this was a draft 😅. Didn't mean to be over-eager. The discord feed didn't show draft at all, so I just clicked it and looked around.

Draft just means I don't want it merged yet, not that I don't want help.

File: worlds/pokemon_rb/client.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Alchav) [reply worlds/pokemon_rb/client.py:204]: > Seems complicated. Can't you just do `money = int(original_money[0].hex())`?

int(original_money.hex()) works. I was not aware .hex() on a bytes object would do this.

File: worlds/pokemon_rb/client.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Zunawe) [reply worlds/pokemon_rb/client.py:244]: `ctx.stored_data` is a dict of all the data your client has requested with `set_notify`. So the energy link value would already be in there and up-to-date.

[the remark this answers is not in the document]
Comment (Silvris) [reply worlds/pokemon_rb/client.py:247]: `ctx.current_energy_link_value` is used by the GUI to display energylink, but that's irrelevant if `ctx.gui.enable_energy_link` is never called.

[the remark this answers is not in the document]
Comment (Alchav) [reply worlds/pokemon_rb/client.py]: Done

File: worlds/pokemon_rb/client.py [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+    "CrashCheck4": (0x16DD, 1),
+}
+
+location_map = {"Rod": {}, "EventFlag": {}, "Missable": {}, "Hidden": {}, "list": {}, "DexSanityFlag": {}}
+location_bytes_bits = {}
+for location in location_data:
+    if location.ram_address is not None:
+        if type(location.ram_address) == list:
+            location_map[type(location.ram_address).__name__][(location.ram_address[0].flag, location.ram_address[1].flag)] = location.address
+            location_bytes_bits[location.address] = [{'byte': location.ram_address[0].byte, 'bit': location.ram_address[0].bit},
```
Comment (Zunawe) [inline worlds/pokemon_rb/client.py:44]: Still a bunch of single-quoted strings here and there that should be double-quoted

**Comment by Berserker66** [comment]
Note that  InnoSetup will not delete old files, like the old .exe or the old .lua unless instructed to do so. I did not see any such instructions. I'd strongly recommend that this is added as users are likely to just try to use the old files if they exist.

**Review by Berserker66 — APPROVED** [review]

**Pull request merged**